care/utils/assetintegration/base.py

```diff
@@ -52,3 +70,119 @@ def api_get(self, url, data=None):
             return response
         except json.decoder.JSONDecodeError:
             return {"error": "Invalid Response"}
+
+    def validate_action(self, action):
+        pass
+
+    def generate_system_users(self, asset_id):
+        asset_queue_key = f"waiting_queue_{asset_id}"
+        if cache.get(asset_queue_key) is None:
+            return []
+        else:
+            queue = cache.get(asset_queue_key)
+            users_array = []
+            for user in queue:
+                users_array.append(User.objects.get(username=user))
+            return users_array
+
+    def generate_notification(self, asset_id):
+        from care.utils.notification_handler import send_webpush
+
+        message = {
+            "type": "MESSAGE",
+            "asset_id": str(asset_id),
+            "status": "success",
+        }
+        user_array = self.generate_system_users(asset_id)
+        for username in user_array:
+            send_webpush(username=username, message=json.dumps(message))
+
+    def add_to_waiting_queue(self, username, asset_id):
+        asset_queue_key = f"waiting_queue_{asset_id}"
+        if cache.get(asset_queue_key) is None:
+            cache.set(asset_queue_key, [username], timeout=None)
+        else:
+            queue = cache.get(asset_queue_key)
+            if username not in queue:
+                queue.append(username)
+                cache.set(asset_queue_key, queue, timeout=None)
+
+    def remove_from_waiting_queue(self, username, asset_id):
+        asset_queue_key = f"waiting_queue_{asset_id}"
+        if cache.get(asset_queue_key) is None:
+            return
+        else:
+            queue = cache.get(asset_queue_key)
+            if username in queue:
+                queue = [x for x in queue if x != username]
+            cache.set(asset_queue_key, queue, timeout=None)
+
+    def unlock_asset(self, username, asset_id):
+        if cache.get(asset_id) is None:
+            self.remove_from_waiting_queue(username, asset_id)
+            self.generate_notification(asset_id)
+            return True
+        elif cache.get(asset_id) == username:
+            cache.delete(asset_id)
+            self.remove_from_waiting_queue(username, asset_id)
+            self.generate_notification(asset_id)
+            return True
+        elif cache.get(asset_id) != username:
+            self.remove_from_waiting_queue(username, asset_id)
+            return False
+        return True
+
+    def lock_asset(self, username, asset_id):
+        if cache.get(asset_id) is None or not cache.get(asset_id):
+            cache.set(asset_id, username, timeout=None)
```

Comment:
having a timeout is a good idea, the correct way would be to call a celery task with some dalay(eg 30 min) that removes lock from the current user

Reply:
In the front-end it is implemented to call the unlock method whenever tab is closed or something